configs: Improve security of in-repo configuration.

matts1 opened the pull request:
Fixes #3303 and #1595

This completely changes the workflow for in-repo configuration. Instead of being stored in a file on disk, we store the configuration wrapped with a signature in order to determine who wrote the configuration.

In the event that the config is considered insecure, we warn the user and require them to run `jj config edit --repo/workspace` to re-enable it (test_config_security.rs simulates these workflows and is relatively well documented).

If applicable:

- [x] I have updated `CHANGELOG.md`
- [x] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [x] I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

matts1 commented:
Notes for reviewers to make your life easier. I recommend reviewing in the following order:
1) Start with `test_config_security.rs` to see the UX
1) Then go to `user_config.proto` then `user_config.rs` to see the API
1) Then look at `cli/config.rs` to view the implementation

Everything else is just mostly boilerplate changes with little substance.

adamse commented on cli/src/config.rs:
```diff
@@ -488,108 +499,148 @@ impl ConfigEnv {
+        let legacy_config = path.join("config.toml");
+        match std::fs::read_to_string(&legacy_config).context(&legacy_config) {
+            Ok(content) => {
+                self.set_repo_config(&content, true)?;
```
it would be nice if migration was noted on stdout/err. less confusion if you don't read the release notes carefully.

adamse commented on cli/src/config.rs:
```diff
@@ -488,108 +499,148 @@ impl ConfigEnv {
     /// Sets the directory for the workspace and the workspace-specific config
     /// file.
-    pub fn reset_workspace_path(&mut self, path: &Path) {
+    pub fn reset_workspace_path(&mut self, ui: &Ui, path: &Path) -> Result<(), CommandError> {
```
if this change doesn't land before workspace config is released there should be a migration like for repo config here.

matts1 replied:
I figured that even if it was released by then it'd be so recently released that my guess is that only you, as the PR author, would probably be the only person who has workspace config.

I can do this if need be, but I'm not convinced it's worth the effort.

matts1 replied:
if it's an option, I'd prefer we temporarily disable workspace configs for the next release if this code hasn't landed by then, to avoid introducing more tech debt. I think it's probably ok to wait one extra release for the per-workspace configs.

I don't have issues with per-workspace configs, just don't want to have to go through the same migration process for them that we do for per-repo configs to migrate to this new format.

yuja replied:
@adamse are you okay with dropping workspace-config.toml entirely and replacing it with `.jj/repo/config.toml` with some `--when` conditionals? Temporary removal wouldn't help much if we're likely to reenable it.

adamse replied:
workspace config has the upside that you don't have to write absolute paths compared to --when.workspaces. to me having the workspace config seems like a useful feature, but i'm not too fussed as my usecases are also covered by when (but i'm not sure about the other people who were pushing for workspace config).

do we think it's likely this pr doesn't land soon? do we think workspace config is not a useful feature?

if workspace config is good and this pr doesn't land soon, we can mention it as experimental in the release notes and require manual migration once this pr lands?

yuja replied:
> workspace config has the upside that you don't have to write absolute paths compared to --when.workspaces.

That could be mitigated by adding `--when.workspace_names`.

> do we think it's likely this pr doesn't land soon?

Probably not. This isn't an easy feature. However, I also think that something like this will have to be added.

PhilipMetzger replied:
> I personally consider that workspace config can be replaced with repo config + conditionals.

I don't think this is true for all the people who want to run their coding agents in separately configured workspaces. Since they want a one-time setup for the agents environment. 

> I don't mean the workspace config is useless, but I don't think it would be worth the effort/complexity to add two separate config + signing file pairs. (It would be nice if we didn't have repo/workspace config files that could be a security risk, but removing all of them wouldn't be acceptable.)

Agreed, having two signed pairs kinda complicates all of this. 

> If we're sure to support workspace config, I think manual migration (without temporary removal) is good enough.

I also agree with that. Although having an automated way would improve the user experience. 

> Another concern is that the priority of repo-managed/repo/workspace configs is a bit weird. Since repo-managed config is actually selected per workspace (or working-copy commit), it could be considered a workspace-level config.

Isn't this technically a bug? Because otherwise this seems to be a really bad oversight.

matts1 replied:
I've actually been having discussions with my team about managed repo config, and they raised some very good points that made me completely reconsider how it should work. Specifically, we have a fake "repo-managed" config currently which is just a symlink to an in-repo config file in chromium.

The issue they've been running into is that they've been doing things like `jj new <LTS branch>` to cherry-pick certain fixes onto said branch. When they do so, because the LTS branch is old, they don't have any configuration. This also results in very confusing UX when one workspace has one version of the jj config checked out while another workspace has a different version. In general, this approach:
* Increases implementation complexity
* Increases complexity from a user's perspective
  * They might get it out of sync whenever they run `jj new`
  * They could have one alias work when you're in one place, then suddenly disappear in a different place.

The approach we think would work better is actually to just always read the repo-managed config from `trunk()`. AFAICT the pros and cons are:
* Pro: *Much* simpler to implement
* Pro: *Much* simpler from a user's perspective
* Pro: Consistent user experience between versions
* Cannot test new configurations easily, since it reads only submitted configs
  * Can be mitigated by adding a global flag to read the config from `@` instead. This would bypass trust, and thus mean that they don't have to review their own configs that they wrote.
* Pro: Any references to out-of-repo tools will actually work with old versions
* Con: Any references to in-repo tools may break if you move the tool
  * The same applies if the tool didn't exist at an old version, but the previous approach also wouldn't solve that problem

AFAICT, the only downside can be relatively easily mitigated by just not moving tools, or moving tools but adding symlinks, etc.

PhilipMetzger commented on cli/tests/common/test_environment.rs:
```diff
@@ -143,6 +143,14 @@ impl TestEnvironment {
+                    .join("Roaming")
+                    .to_str()
+                    .unwrap(),
+            );
```
nit: probably should move to a separate commit with its own explanation of why it was done

PhilipMetzger commented on lib/src/user_config.rs:
```diff
@@ -0,0 +1,243 @@
+    fn filename() -> &'static str {
+        "workspace-config.binarypb"
+    }
+
```
nit: these should end with `binpb` per Google's (resp: the protobuf teams) own suggestion in https://protobuf.dev/programming-guides/techniques/

PhilipMetzger commented on lib/src/user_config.rs:
```diff
@@ -0,0 +1,243 @@
+    Ok(match std::fs::read(&signing_key).context(&signing_key) {
+        Ok(bytes) => SigningKey::from_pkcs8_der(&bytes)?,
+        Err(e) if e.source.kind() == std::io::ErrorKind::NotFound => {
+            std::fs::create_dir_all(&key_dir).context(&key_dir)?;
```
nit: simplify the names here by importing a bunch of stuff

PhilipMetzger commented on lib/src/user_config.rs:
```diff
@@ -0,0 +1,243 @@
+    let buf = match std::fs::read(&path).context(&path) {
+        Ok(buf) => buf,
+        Err(e) if e.source.kind() == std::io::ErrorKind::NotFound => {
+            if d.exists() {
```
ditto with imports

PhilipMetzger commented on lib/src/user_config.rs:
```diff
@@ -0,0 +1,243 @@
+}
+
+/// Reads the user configuration from a repository for the current user.
+pub fn read_user_config<T: ConfigType>(d: &Path) -> Result<UserConfig<T>, UserConfigError> {
```
nit: `d` isn't really a descriptive variable name

matts1 commented on CHANGELOG.md:
```diff
@@ -66,6 +66,11 @@ to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
   tags. These keywords may be useful in non-colocated Git repositories where
```
I'll leave this open, depends on whether we land this before a new release.

PhilipMetzger commented on cli/src/commands/git/mod.rs:
nit: squash this and the nix commit into the first one (although you may need a Nix expert to help you with the problems there before doing that).

yuja commented on lib/src/user_config.rs:
```diff
@@ -0,0 +1,245 @@
+    };
+    let key_dir = strategy.data_dir().join("jj/config/keys");
+    let signing_key = key_dir.join("ed25519.der");
+    let verifying_key = key_dir.join("ed25519.der.pub");
```
I don't think we need public-key cryptography to sign and verify user's private data by his key. What we need is something like `hash(private_key + path + config_data)` but is considered secure. I _think_ some MAC algorithm can be used.
https://en.wikipedia.org/wiki/Message_authentication_code

I'm not an expert, but pubkey-based algorithm would be computationally more expensive than symmetric-key algorithms or cryptographic hashing.

matts1 replied:
I double-checked with a security expert, they say that you're right and that asymmetric keys aren't improving anything.

They also mentioned that specifically, "HMAC_SHA256 would be a sensible choice"

matts1 commented on cli/src/config.rs:
```diff
@@ -488,108 +499,147 @@ impl ConfigEnv {
+        match std::fs::read_to_string(&legacy_config).context(&legacy_config) {
+            Ok(content) => {
+                self.set_repo_config(&content, true)?;
+                std::fs::remove_file(&legacy_config).context(&legacy_config)?;
```
> I don't quite see how you have a config file in the repo without having a file.

You have a file, it's just that the file, roughly speaking, instead of being a string representing the configuration, contains something more akin to the struct:
```
struct SecureConfig {
  config: String,
  // sign(config, my credentials)
  signature: Vec<u8>
}
```

So it is stored in a file, it's just that the file isn't directly edited via a text editor, any modifications are wrapped by `jj config edit`

> Wouldn't all the commands that deal with files have to special case it?

No, because this isn't stored in the repo. No commands read / write the config file (other than `jj config`, which I've already done in this PR)

yuja replied:
> > Deleting `config.toml` would break forward compatibility, so it would require grace period.
> 
> I have already done this. It automatically migrates from the old config to the new one.

and deletes the old config file, right? It would break old `jj` versions and external tools compiled against old jj-cli/jj-lib.

> You can run `jj config edit --repo` to review someone else's repository config without trusting it (since their repo config won't be enabled initially). And if you don't trust it, you delete the parts you don't like (or the whole thing) so it's safe.

Doesn't we need to make the editor quit with non-zero exit code or something to not trust it? Also, we can't use other tools (e.g. grep, diff, etc.) without materializing the content, which would be inconvenient.

I would just add signature file alongside `config.toml` because that would be simple and forward compatible. Old `jj` can read `config.toml` ignoring trusts.

matts1 replied:
Ok, done. On the upside, the code is a lot simpler now because of that choice.

yuja commented on cli/src/config.rs:
```diff
@@ -532,8 +656,11 @@ impl ConfigEnv {
+            self.load_user_config("repo", &path.parent().unwrap().join(REPO_CONFIG_FILE), path)?;
+        }
         if let Some(path) = self.existing_repo_config_path() {
             config.as_mut().load_file(ConfigSource::Repo, path)?;
```
Oh, I meant the config content would still be verified.

```
let data = read("config.toml");
let signature = read_and_decode("config.sig");
verify(&path, &data, &signature)
```

@PhilipMetzger what was your idea?

matts1 replied:
Doing so would significanty degrade the user experience. Whenever I manually edited `.jj/repo/config.toml` in an IDE instead of via `jj config edit --repo` I would have to reapprove the changes.

matts1 replied:
I wrote a design doc for this feature, so we can discuss this there https://docs.google.com/document/d/1zopSHcb8BISsnf9SeT541fO_-XgY0Kvei128h5uO-tw/edit?tab=t.0

matts1 replied:
The original PR didn't have this issue because you could only edit the config via `jj config edit`, so the signature could never get out of sync with the config.

If you store the trusted paths outside of a repository, such as `~/.local`, then the trust is associated with a path, not a repository. This now introduces a new attack vector of:

1) "please help, here's a zip file" (it's a safe one)
2) I download it, verify that it's safe (it is)
3) "wait no, I sent the wrong zip file" (sends a malicious one)
4) The user is almost certainly going to delete the zip file and replace it with a new one at the same location.

yuja replied:
> The original PR didn't have this issue because you could only edit the config via `jj config edit`, so the signature could never get out of sync with the config.

True, and we can now edit config file without using `jj config edit` and re-approve it. That isn't worse than the original PR, I think.

> If you store the trusted paths outside of a repository, such as `~/.local`, then the trust is associated with a path, not a repository. This now introduces a new attack vector of:
> 
> 1. "please help, here's a zip file" (it's a safe one)
> 2. I download it, verify that it's safe (it is)
> 3. "wait no, I sent the wrong zip file" (sends a malicious one)
> 4. The user is almost certainly going to delete the zip file and replace it with a new one at the same location.

Since the signature is still associated with the path, a malicious user could ask me to send back the repository between the step 2 and 3. I agree this would be more secure than simple path-based trusts, but the user would have to be careful to not leak the path-based signature.

FWIW, I personally wouldn't want to enable repo config in someone else's workspace.

PhilipMetzger replied:
> @PhilipMetzger what was your idea?

What do you mean?

PhilipMetzger replied:
I think signing it by both path and content makes sense in the long-term. But we always can start out just with the path.

yuja commented on lib/src/protos/user_config.proto:
```diff
@@ -0,0 +1,43 @@
+  // same signature, and I would then have user configs enabled, despite the
+  // repo coming from someone else.
+  // We mitigate this risk by putting the path in as part of the signature.
+  string path = 3;
```
This is an idea for future extension, but it's nice if we can trust config for path prefix such as `/home/user`. For example, suppose the user doesn't unzip untrusted repositories under `~/work`, they can set a config knob to sign their repo config files for `~/work`. Then they can move their repositories without re-signing configs.

matts1 replied:
Keeping in mind the fact that this only occurs precisely when you move a repository that contains a custom `.jj/repo/config.toml`, this should be an extremely rare occurence.

I'm not a fan of trying to make optimizations to minimize the impact of this.

matts1 commented on cli/src/cli_util.rs:
```diff
@@ -3858,11 +3858,45 @@ impl<'a> CliRunner<'a> {
+        // will return an error telling you to run `jj config edit`.
+        // Hence, `jj config edit` needs to be able to run even when reloading
+        // the config would return an error.
+        let allow_in_repo_config = if let Ok(m) =
```
Roughly speaking, our options are:
* Load without warning (what we currently do)
* Warn + load
* Warn + don't load
* Error

There are only a few circumstances where this feature can pop up, specifically:
* You have moved a repo containing a `.jj/repo/config.toml` file
* You now have a different user account (eg. copy to a new computer without copying the whole home directory)
* Someone else has sent you a zip file of a repo which contains a `.jj/repo/config.toml` file

In the first and second situation, *you* wrote the config files yourself.
* Load without warning: This would be ideal, but since this is unable to be distinguished from the scary option, it is not feasible
* Warn + load: Every time you run jj, you're going to see a warning. The user will see annoying messages until they follow the instructions and approve it
* Warn + don't load: Every time you run jj, you're going to see a warning. The user will see annoying messages until they eventually approve it. Also, the config they wrote won't work
* Error: They see an error once, with clear instructions on how to approve it. They run the command and it's immediately solved.

In the other scenario, you're using someone else's config:
* Load without warning: This would be terrible. Your system will likely get compromised
* Warn + load: Also terrible, by the time you get warned your system may already be compromised
* Warn + don't load: This is all right from a security standpoint. But:
  * Every time you run jj, you're going to see a warning.
  * It's highly likely that you're debugging an issue that that user had, in which case you really should be trying to enable their repo-specific config.
* Error: JJ refuses to run until you resolve this. Either:
  * The user doesn't want the config, in which case they run `jj config review` and delete all lines
  * The user does want the config, in which case they run `jj config review`, save and quit
  * The user does want the config, but doesn't like a specific line of it, so they run `jj config review`, deletes that line, then saves and quits

IMO, we should always error, because loading is insecure, warning + don't load is secure but provides a bad user experience.

In what specific ways are you concerned about this feature impacting users? I don't think this feature needs a gradual rollout, as I don't believe it actually affects any user journeys, and even if it did, all it results in is about 10 seconds of time wasted as the user re-approves it, since the error provides clear instructions on how to fix it.

yuja replied:
> > I don't think this is uncommon, and that's one reason I said we would have to try this feature out without enforcing.
> 
> I think that as long as we can make the UX good, it should be fine to enable. If people take issue with it and we get complaints, we can reconsider it then. I think that a gradual rollout means that you just won't get the feedback you need on whether or not it's the right solution.

To get a feedback without breaking things, I think "warn + load" is needed. And if the user gets annoyed, the warning can be turned off until we improve the UX.

It's not obvious how often user renames repository, for example. You said it's rare, but I occasionally do (a couple of times in a year or something.) If config verification is enabled with the current form, a simple `jj git clone URL && mv x y` would make the repository unusable. I'm afraid there would be other UX problems which I'm not aware of.

> > It can be waste of time to review the config which I'm not interested in. Right now, I have to manually check if there's `.jj/repo/config.toml`.
> 
> This feels like something that can relatively easily be addressed with some better UX instead. How about I do what I was planning to do for repo config early. I'll introduce three new commands you can run:
> 
> * `jj config review --repo`: Basically just an alias for `jj config edit`
> * `jj config review --repo --trust`: Marks a config as trusted without having to review it
> * `jj config review --repo --distrust`: Marks a config as untrusted without having to review it

These commands might be useful, but I don't think it's way better than just emitting a warning.

matts1 replied:
I think I've found a good solution to this. It now does the following instead:
```
$ jj <command>
Warning: The repo has moved from $TEST_ENV/repo/.jj/repo to $TEST_ENV/copy/.jj/repo. For security reasons, we don't trust your repo config
There are several things you can do with this config:
  (r): Review the config and make any changes required to make it safe
  (t): Trust the config
  (d): Delete the config
Please select an option: t
<command proceeds, config is trusted>
```

So you can now approve the config in two keystrokes, this should have a minimal impact on users.

matts1 replied:
I don't think any option is going to be a "good" UX. No matter what we do, the UX will inherently make things more annoying for the user. So we've got to accept some level of that. It's simply a question of the least annoying option.

Unexpected interaction occurs on one side, but each other option provides its own pitfall:
* Warn + load is insecure, and spams the user on every jj invocation
* Warn + don't load spams the user on every jj invocation
* Warn once + don't load means that the user has no clue why their in-repo config isn't loading
* Error means that a command that the user thought "couldn't fail" will fail

The advantages of this approach is that:
* This only needs to be done once (after that it's resolved)
  * This is super important, as every approach involving warnings would frustrate me like crazy because it would happen on every jj invocation.
* This can be resolved in two keystrokes, so is hence the easiest and fastest to resolve
  * This option is, by far, the least time consuming option for the user
* Instead of giving the user one option (`jj config edit --repo`), we can easily give the user three options (review/trust/distrust), as we don't need to be anywhere near as verbose.
  * This allows for the really streamlined "trust" and "distrust" options

I don't think anybody "likes" unexpected interaction, as such, but I think that it will have less annoyance than any of the alternatives. My personal feelings is that I would get annoyed that I'm pulled out of my expectations, but I would appreciate that it gave me several resolutions that I could easily choose between instead of just erroring out "no" or spamming me with warnings.

So TLDR, I think that it's not perfect, but it's the best we can do. I also think that this is one of those things where we can always submit it now and change between any of these options later. If we get a bunch of feedback and decide that we want to do another option, it's not difficult to change.

yuja replied:
> I don't think any option is going to be a "good" UX.

Just to be clear, my original suggestion of "warn+load"/"load without warning" was for dogfooding of the current signature-based solution. I'm not saying that the current design will work out if we have a slightly better UI. My concern is that it could be annoying enough to be disabled by default until we find a better workaround.

---

For typical users who just clone repositories, it's probably good to trust everything under their home/project directory. This wouldn't require config/path signatures and `mv` just works (so long as the directory isn't moved across trust boundary.) One of the main advantages of using signatures is/was that the trust isn't associated with paths but contents. Unfortunately, as you mentioned, it doesn't always work because a trusted config file in a certain repository might not always be okay in other repositories. So, I don't know if it's good compromise to use signature to sign only repository paths.

matts1 replied:
> Just to be clear, my original suggestion of "warn+load"/"load without warning" was for dogfooding of the current signature-based solution.

What do we get out of this though? The point of dogfooding is that you can try something out before committing to it, but in this case, the thing you want to try out is different to the final solution proposed (and you'll get plenty of people like me who would hate warn + load, but be perfectly fine with the current solution, so you can't really use it to test the waters either)

> My concern is that it could be annoying enough to disable by default until we find a better workaround.

I struggle to see an argument that it's that much of an annoyance. This can only occur as frequently as a user moves their repos, and so if a user runs `mv foo bar`, adding two keystrokes to what's already 10 keystrokes could hardly be described as anything more than an extremely minor annoyance, IMO, certainly not enough to justify disabling by default. Additionally, if for whatever reason it does happen to be annoying enough that we need to disable it by default, we can just do that until we come up with something better.

I don't think that the decision on the precise UX needs to be final (responding to user feedback is important), but I do think that it needs to be what you intend to release to the end users. If we release warn + load we won't get feedback on whether this was a good idea - we'll get feedback on whether warn + load was a good idea. That being said, I think that it is important that we put our best foot forward and come up with the solution that seems most likely to get positive feedback (or in the case of a security change, the least negative feedback). I believe this is that solution.

> For typical users who just clone repositories, it's probably good to trust everything under their home/project directory. This wouldn't require config/path signatures and `mv` just works (so long as the directory isn't moved across trust boundary.)

I'd go a step further and say that for the kind of user that never downloads a zip file, rather than trusting specific directories, they should disable this feature entirely. I'd say that for this kind of user, their ideal scenario is that we completely disable this feature.

However, I don't think that a toggle needs to exist in the initial version, as disabling (or weakening if we just allowlisted specific directories) security features IMO should require a better understanding of what the pain points are that we'll only get from seeing the security features in action. My personal opinion is that the pain of this feature is sufficiently minor and infrequent that it just isn't worth the security risk of providing a toggle, but we'll need to see how users actually feel about this in practice to determine whether this is actually true.

> One of the main advantages of using signatures is/was that the trust isn't associated with paths but contents. Unfortunately, as you mentioned, it doesn't always work because a trusted config file in a certain repository might not always be okay in other repositories. So, I don't know if it's good compromise to use signature to sign only repository paths.

My [design doc](https://docs.google.com/document/d/1zopSHcb8BISsnf9SeT541fO_-XgY0Kvei128h5uO-tw/edit?tab=t.0#heading=h.lldt5vlxjhnx) details precisely what we lose out on. Specifically, CUJ 3 requires signing the config to deal with it correctly.

I think claiming that "it may not be a good compromise" is entirely fair, but this PR is relatively decoupled from whether we sign the config itself. I purposefully introduced an abstract mechanism through which you can sign arbitrary metadata about the repository, and this means that if we decide that it's not good enough in the future, we can just add a string to the protobuf and it'll "just work". I think that for that reason, I agree with @PhilipMetzger that it's ok to submit this as-is and then we can decide to change that down the line if we want.

matts1 replied:
This discussion seems to be going nowhere, it feels to me like we've both been reiterating the same points several times. There are three questions at play here:
1) What security model should we use
1) What should the UX be
1) Should we be providing some kind of opt-in or opt-out

 I've been getting rather frustrated by your review since it feels like you're making complaints about tradeoffs I've made without making a decision on that tradeoff. For example, the feedback of "I don't like the UX of signing the repo paths" which you keep iterating isn't really useful. I don't like it either, but it improves security, and I'm happy to not sign the repo paths. What I need is either:
* "I don't like the UX of signing the repo paths, but I think it's worth the security benefits. LGTM"
* "I don't like the UX of signing repo paths, and I don't think it's worth the security benefits. We should get rid of it"
* "I don't like the UX of signing repo paths, and I have a different idea on how we can resolve the security issues signing repo paths was designed to address".

Re the security model, I'm honestly not particularly fussed, and I can go with whatever you think as long as the UX works for it. I'd appreciate it if any further discussions about the security model were done in comments on the design doc though, as I think that google docs provides a much better platform for back and forth discussion.

Signing the configs also has a few different options:
* My initial implementation involved signing configs, but the compromise was that you couldn't edit configs except via `jj config edit`.
* We could keep storing it on disk, but we have to then acknowledge that the configs will get out of sync if you manually edit the config file
* The current implementation is to not sign the configs, but obviously that's not as secure.

Similar to above, it feels you've been complaining about wanting the config to still be verified, I said that would have consequences (config going out of sync), and I haven't yet received a response regarding those consequences. I need you to either accept one of the above options, or present another option.

Re what should the UX be, I have a few constraints I want to work under:
* An untrusted config cannot, under any circumstances, be loaded
* If there is a user to ask, we should give them options about what to do in a prompt
  * This provides the best time-to-resolve this issue, and minimal UX overhead
* If there's not a user to ask, I would prefer to error out but can compromise on this (as long as we don't load the config).

Re opt-ins / opt-outs:
* Opt-ins are pointless. No-one will actually enable this feature, so it won't get us anything
* Opt-in with a warning means you knew a config was insecure but you loaded it anyway, so I refuse to budge on not implementing this
* I won't implement an auto opt-out for specific subsets (eg. the home directory as you suggested earlier)
  * While *you* might never download something to your home directory, that's something that jj should definitely not assume. For example, I don't ever download zip files of repos, but if I did, I would store it in my home directory, and most people would probably store it in their downloads folder in their home directory. 
* I strongly believe that this feature is nowhere near as debilitating for the user as you seem to make it out to be, and that users will not care. I would much prefer not to implement an opt-out for this initial version of the feature. However, I am willing to compromise on this if that's what's needed to get your approval.
  * I'm not a fan of your specific directory opt-out idea, but it could be achieved via a `--when` clause on the opt-out
 
Are the above constraints reasonable? Any you take issues with? And if not, what would you propose, from a UX perspective? And what tradeoffs would this be making in terms of security vs UX?

> Since the protobuf data can be controlled by an attacker, I don't think we can freely extend it to make it more secure. If new `jj` version add `config_content_hash: Vec<u8>` field for example, old proto data which doesn't have this field cannot be trusted. We can of course add migration period, though.

Correct, but that's no different to what we're currently doing with a migration period to add this metadata. My point was that it's not a difficult change to make.

> EDIT: and we might have to generate old/new signatures so old `jj` which doesn't know the new field can verify it. (or sign the serialized proto data.)

We won't need to do that. We already sign the serialized proto data.

> Often, there is no user there to ask. I have lots of repository uses via scripts, these could clone or create+destroy workspaces, e.g. for building release assets, parallelizing CI, docker builds, etc.

Yes, my current implementation takes that into account and checks `Ui::can_prompt()` to see if there is a user to ask. If there isn't, it throws an error instead of asking the user to resolve it.

yuja replied:
I'm also getting tired of reviewing this and related PRs, so  better to ask someone else for opinions. I hope someone else will take over this, and the review of design docs, which I'm really bad at.

I'm asking to add escape hatch and try this out as an opt-in/opt-out feature (repeatedly as you said) because I can't decide if the current model would be accepted by users. (I personally don't think it would, but it's just my feeling, and you apparently have different opinion.)

> We won't need to do that. We already sign the serialized proto data.

That's good to know. I thought signature was computed from values, but maybe that was wrong.

matts1 replied:
I agree with everything you just said, except github is a pain in the ass to comment and discuss on, so my plan was to do the design review in google docs via comments and then convert it to markdown once it's been approved.

matts1 commented on lib/src/protos/user_config.proto:
```diff
@@ -0,0 +1,43 @@
+  string path = 3;
+
+  // sign(storage + salt + path)
+  bytes signature = 4;
```
Still not sure how it works, but I implemented it regardless

yuja replied:
I'm not familiar with Windows either, but the situation would be similar to NFS. A directory may be mounted at different location on different machines.

matts1 commented:
@yuja There's clearly been a lot of contention with the UX issues, so I've been rethinking from the ground up, and realized that if I added a concept of a "repo ID" and "workspace ID", we could store configs out-of-repo and solve almost all the issues we were having, and not require any kind of user intervention, or disabling of configs. I've managed to mitigate most of the concerns I originally had with out-of-repo configs, and there's only some very minor UX wierdness.

I wrote the design doc after writing the initial PR, but this time I've updated the [design doc](https://docs.google.com/document/d/1zopSHcb8BISsnf9SeT541fO_-XgY0Kvei128h5uO-tw/edit?tab=t.0) and haven't yet written any code. I think you'll appreciate the new design, I put a lot of effort into explaining the solution in an easy-to-understand manner, and I think you'll find it much easier if we do it in a two step process where we first review the design doc and then the PR, so I'd like to invite you to take a look at the new design and review it, and then after it's approved I'll start work on the PR. If you'd still prefer, I can find a different reviewer, but I think you'll find that the contention has disappeared, as I think this new approach gives you what you want out of this.
